### PR TITLE
{folly,fizz,mvfst,wangle,fbthrift,fb303,edencommon,watchman}: 2024.03.11.00 -> 2024.11.18.00, refactor

### LINKED ISSUE #334685: Update request: watchman 2024.03.11.00 → 2024.08.12.00
- Package name: watchman
- Latest released version: 2024.08.12.00

- Current version on the unstable channel: 2024.03.11.00
- Current version on the stable/release channel: 2024.03.11.00

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**
@kylesferrazza @sdht0

-----

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by emilazy**
This was no fun at all.

The primary impetus here was to get LLVM 19 fixes, but I ended up doing way more clean‐up and improvements than I intended. I’d recommend reviewing this commit by commit, but… there’s so much of it and it’s mostly the same stuff over and over. It might be best to just review these as new derivations. I’ve added myself as maintainer to all of them, although I only care about the libraries insofar as Watchman depends on them.

I’m not proud of some of the things I’ve done here and there’s definitely quite a lot of duplication that could use factoring out, but it seems like Facebook is in the process of moving these builds to Buck2, which I’m sure will make them an even more enjoyable experience to package, and would anyway likely obsolete a lot of what there is to be factored out. The header thing especially should probably be a hook or done in the compilers or something, but I’m tired and right now I just want this saga to be over with. Oh, and some of the new tests that aren’t yet disabled are probably flaky too, but I’ll deal with them as they come up on Hydra…

@techknowlogick I see you started working on this in https://github.com/NixOS/nixpkgs/pull/347352. Do you want to help maintain these packages with me? I’m in dependency hell and I fear I may never be able to crawl out.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/mv/mvfst/package.nix
Comment (paparodeo) [inline pkgs/by-name/mv/mvfst/package.nix:100]: this look like globs and not a regex.

File: pkgs/by-name/mv/mvfst/package.nix
Comment (paparodeo) [inline pkgs/by-name/mv/mvfst/package.nix:79]: this is on x86 darwin only? mvfst built for me on aarch64 tho don't recall if I was at the latest rev, maybe late October.

File: pkgs/by-name/mv/mvfst/package.nix
Comment (emilazy) [reply pkgs/by-name/mv/mvfst/package.nix:100]: Those are the actual test names. The regex is escaped so it matches precisely. I don’t know why it’s like that, either.

File: pkgs/by-name/mv/mvfst/package.nix
Comment (emilazy) [reply pkgs/by-name/mv/mvfst/package.nix:79]: I think a combination of updating + building shared libraries breaks it. Homebrew sets it unconditionally: https://github.com/Homebrew/homebrew-core/commit/7307d8bf53e9c4100e9f52bca6191fb6f660aff8.

File: pkgs/by-name/mv/mvfst/package.nix
Comment (paparodeo) [reply pkgs/by-name/mv/mvfst/package.nix:100]: oh yeah. I missed the escaped regex function.

**Review by paparodeo — APPROVED** [review]

**Comment by techknowlogick** [comment]
> @techknowlogick ... Do you want to help maintain these packages with me?

yes please :) feel free to add me as a maintainer in this PR, or I can do it in a different one once this is merged.

Thanks so much for going through all of this!

File: pkgs/by-name/mv/mvfst/package.nix
Comment (emilazy) [reply pkgs/by-name/mv/mvfst/package.nix:79]: I’ve since learned that although this option is deprecated, it doesn’t force a flat namespace, and it basically just makes `ld64` behave more like the GNU linker with regards to undefined symbols. An upstream fix to add the necessary libraries would be ideal, but the final executables are at least guaranteed to not have anything unbound.

**Comment by techknowlogick** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356219`

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>buck</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 23 packages built:</summary>
  <ul>
    <li>cpptoml</li>
    <li>edencommon</li>
    <li>edencommon.dev</li>
    <li>fb303</li>
    <li>fb303.dev</li>
    <li>fbthrift</li>
    <li>fbthrift.lib</li>
    <li>fizz</li>
    <li>fizz.bin</li>
    <li>fizz.dev</li>
    <li>folly</li>
    <li>folly.dev</li>
    <li>mvfst</li>
    <li>mvfst.bin</li>
    <li>mvfst.dev</li>
    <li>python311Packages.pywatchman</li>
    <li>python311Packages.pywatchman.dist</li>
    <li>python312Packages.pywatchman</li>
    <li>python312Packages.pywatchman.dist</li>
    <li>rs-git-fsmonitor</li>
    <li>wangle</li>
    <li>wangle.dev</li>
    <li>watchman</li>
  </ul>
</details>

**Comment by techknowlogick** [comment]
> Great! Added you as a maintainer here, no other changes :)

Thanks!! Took a sec to build, but looks like everything is passing. Think this is good to merge in now :)

**Review by techknowlogick — APPROVED** [review]

**Comment by emilazy** [comment]
> Took a sec to build, but looks like everything is passing.

You should see how many times I had to rebuild this whole stack when getting it all to work properly… 🫠

Updated to the latest version; will probably merge in a day or two so I don’t have to contemplate the prospect of them putting out *another* release.

**Pull request merged**
